### DISCUSSION TITLE
DEV SESSION 11 NOV --- 2000 UTC+8/1400 CET/1200 GMT/ 0600 EST

### DISCUSSION THREAD

**Discussion by tpike3** [mesa/mesa#1849] [General]
Where: [via Matrix](https://matrix.to/#/#project-mesa:matrix.org)

Proposed Agenda:

-Small Development Grant and Visualization Improvements
-Mesa-Geo 
-PRs and Issues

@rht , @jackiekazil , @Corvince , @wang-boyu , @ankitk50, @rlskoeser

**Comment by tpike3** [comment]
1st apologies for messing up the time as the US east coast switch to daylight savings time. (I will get it right in December) 

Below are the notes: 
- We are going to do a Mesa 2.1.5 release this week once we merge #1866 and possibly convert the README to markdown as we continue to convert all .rst files to md. The primary reason for this is @rlskoeser identified that under the new construct the plots are flickering as the render each time. and @rht edit with #1867 fixed this. 
- AAG mesa and mesa-geo submission is complete for April conference. 
- The Mesa community is growing we added several more people to the distro list after CSSSA and the last release. 
- As we look at the front end @ankitk50 is working on building tests for the new front end and doing some additional cleanup. As we look at additional features we are thinking the ability to zoom in (imagine a plot with lots and lots of agents with a similar feature to a typical GIS zoom ability) and the ability to scroll over to get information about  specific agents. The question for the team (specifically @rht, @Corvince  and @wang-boyu ) Should we try and do this organic with matplotlib, or use another library like mpl-interactions, plotly, or bokeh? Please let us know your thought in this discussion.

**Reply by wang-boyu** [reply]
> 1st apologies for messing up the time as the US east coast switch to daylight savings time. (I will get it right in December)

Oh no the fault is mine - I should've kept track of that. Sorry.

**Comment by wang-boyu** [comment]
Regarding interactive plotting, Solara already has support for Plotly...? If users can extract agents as dataframe, they should be able to do something like this: https://github.com/wang-boyu/mesa-examples/blob/gis%2Fsolara-frontend/examples/schelling/app.ipynb

There're some small issues with the code above, but it's also inline with what we discussed here https://github.com/projectmesa/mesa-geo/issues/157#issuecomment-1794476285

**Reply by rht** [reply]
The current choice of using Matplotlib instead of Altair (which is lightweight) is constrained by having to support the graph visualization. But there is a workaround for this: use a plotting library that supports interactivity for other visualizations, while for graph visualization, fallback to Matplotlib. This means Matplotlib and NetworkX would be an optional dependency so as not to bloat the dependencies.

Plotly supports graph visualization. The only downside is that it has huge dependencies.

For comparison:
- `pip install solara altair`: 343 MiB (`pandas` 75 MiB, `numpy` 41 MiB, `numpy.libs` 36 MiB, `debugpy` 23 MiB, `solara` 18 MiB, ..., `altair` 7 MiB)
- `pip install solara matplotlib pandas`: 424 MiB (`pandas` 75 MiB, `numpy` 41 MiB, `matplotlib` 41 MiB, `numpy.libs` 36 MiB, `solara` 18 MiB, ...)
- `pip install solara plotly pandas`: 500 MiB (`plotly` 155 MiB, `pandas` 75 MiB, `numpy` 41 MiB, `numpy.libs`, `debugpy` 23 MiB, `solara`, ...)

Extra note that `solara`'s size has progressively become smaller over new releases.

**Reply by rht** [reply]
Though, the install size should be a secondary concern.
According to the accepted answer in https://stackoverflow.com/questions/59845407/plotly-express-vs-altair-vega-lite-for-interactive-plots
> Plotly currently supports more many more types of graphs and has some special functionality targeted to for example biological plots and image analysis.
> [Plotly can accelerate performance with WebGL](https://plotly.com/python/webgl-vs-svg/) for certain types of plots, whereas [Altair's performance can be scaled with VegaFusion](https://vegafusion.io/#).

**Reply by tpike3** [reply]
That is a great roll-up, my thought right now would be plotly, based in its syntax, greater number of graphics and Solara integration.  

However, @rlskoeser , @jackiekazil  and @ankitk50 what are your thoughts?

**Reply by ankitk50** [reply]
Both Altair and Plotly are supported by solara. Altair is lightweight and supports most of the interactive features like tooltips, clicking,  that Plotly does. The heavy lifting is done in syntax for Altair and package size for Plotly. 

But if we are okay with package size, I would also choose Plotly.

**Reply by rlskoeser** [reply]
I've been using altair for my mesa project because I'm more familiar with it and it solved a problem with plotting my agent space for something I couldn't easily do in matplotlib (I think it was setting colors based on a data attribute, but I may be misremembering). I don't have any particular objection to plotly. Would be good to keep it possible for people working with mesa to still use altair if they want to, since solara supports both.

**Reply by tpike3** [reply]
Based on on this it seem like altair is the way to go. Any disagreements?

**Reply by rht** [reply]
According to https://github.com/vega/vega/issues/1738, Vega has 3D visualization, but this has yet to be wrapped in Altair.
My take would be to go with Solara's route, with several configurable plotting backends, but defaults to Altair. When users want to display network space or 3D space, we should raise an error message that says that they need to select the Plotly backend, and to have Plotly installed. This way, `pip install mesa`, by default, doesn't take much space, just like in `pip install solara`. This should enable running the visualization code on Binder.

**Reply by rht** [reply]
> This should enable running the visualization code on Binder.

This has yet to be tested. I wasn't able to recall why the current `JupyterViz` can't run on Binder. It might have something to do with resource limitation.

**Reply by jackiekazil** [reply]
@tpike3 agree to altair.
